NeoML/src/Dnn/Layers/FullyConnectedLayer.cpp

```diff
@@ -22,194 +22,260 @@ limitations under the License.
 namespace NeoML {
 
 CFullyConnectedLayer::CFullyConnectedLayer( IMathEngine& mathEngine, const char* name ) :
-	CBaseLayer( mathEngine, name == nullptr ? "CCnnFullyConnectedLayer" : name, true ),
-	numberOfElements(0),
-	isZeroFreeTerm(false)
+	CBaseLayer( mathEngine, name == nullptr ? "CCnnFullyConnectedLayer" : name, /*isLearnable*/true ),
+	numberOfElements( 0 ),
+	isZeroFreeTerm( false )
 {
 	paramBlobs.SetSize(2);
 }
 
-CFullyConnectedLayer::~CFullyConnectedLayer()
+CFullyConnectedLayer::~CFullyConnectedLayer() = default;
+
+const CSmallMatricesMultiplyDesc* CFullyConnectedLayer::initSmallMatricesMulDescs( TSMMD type, int inputNumber,
+	int firstHeight, int firstWidth, int secondWidth, int resultWidth )
+{
+	NeoPresume( inputBlobs[0] != nullptr || inputDiffBlobs[0] != nullptr );
+	NeoPresume( outputBlobs[0] != nullptr || outputDiffBlobs[0] != nullptr );
+
+	CSmallMatricesMultiplyDesc* desc = smallMatricesMulDescs[type][inputNumber];
+	if( desc == nullptr ) {
+		desc = MathEngine().InitSmallMatricesMultiplyDesc(
```

Comment:
I didn't think through about one of the aspectrs - how to avoid many-many empty calls to `InitSmallMatricesMultiplyDesc` when math engine is not CPU+x86 (e.g. CUDA or CPU+ARM).

What do you think about always return not `nullptr` from `InitSmallMatricesMultiplyDesc`? For example `return new CSmallMatricesMultiplyDesc()` (which is empty class).

Reply:
Seems good, no `CPair` copies all over the code. No empty calls for GPU/ARM math engine... The `CPair` logic is hidden inside `CCpuMathEngine+MKL`.

Reply:
Yes, now it is fixed this way)